Laravel authentication (#3038)
oktadeveloper/okta.github.io: _source/_posts/2019-09-05-laravel-authentication.md

---
layout: blog_post

> title: "Build a Simple Laravel App with Authentication"

Reviewer: 64 on https://brentisarulebreaker.dev/, not too shabby!

author: krasimir

> description: "Let's build a simple Laravel app and add authentication with a custom identity provider"

Reviewer: Add a period at the end of the description.
```suggestion
description: "Let's build a simple Laravel app and add authentication with a custom identity provider."
```

In this article, I'll show you how to get started with Laravel and quickly build an application with user authentication (using Okta as your identity management provider). 

<div markdown="1" style="border: 1px #eedeaa solid; background: #ffffee; padding: 20px; border-radius: 8px;">
**Prerequisites:** Please ensure you have already installed PHP, Composer, and have signed up for a [free Okta developer account](https://developer.okta.com/signup/).
</div>

## Understand the Laravel Model-View-Controller (MVC) Architecture and Middlewares

Laravel's architecture is based on the Model-View-Controller (MVC) pattern. MVC emerged from research of the development of graphical user interfaces and it's not particularly suited to Web applications which are based on a request-response cycle with a noticeable latency and a strict delineation between the frontend and backend. There are certain shortcomings of the pattern when applied to the Web:

* there is some debate in the community about the proper way to handle data validation, user authentication/authorization, session management and other specifics of a Web application in a traditional MVC model;

> * the MVC model needs to be expanded (usually with an additional Service layer) if you want to separate the business logic of the application cleanly while maintaining lean controllers and pure models.

Reviewer: It seems weird that these sentences don't start with a capital letter, yet they end with punctuation.

Laravel handles user authentication by default through its 'auth' middleware. When you start a new application, Laravel provides a way to include boilerplate code for user registration/login/forgot password functionality, plus the authentication middleware and an easy way to protect routes so they're only available to logged-in users.

## Add Identity Management to Laravel

Okta is an API Identity service that allows you to create, edit, and securely store user accounts and user account data, and connect them with one or more applications. We might be biased, but we think Okta makes [identity management](https://developer.okta.com/product/user-management/) easier, more secure, and more scalable than what you're used to.

To complete this tutorial, you'll need to [register for a forever-free developer account](https://developer.okta.com/signup/). 

> You'll replace some of the Laravel boilerplate code in your application so it uses Okta to authenticate users. The Socialite package for Laravel will help make it easier - it can handle integration with a large number of social login providers (including Octa) out-of-the-box.

Reviewer: Octa -> Okta

Author: how did neither brent or I catch this? 😂

## Laravel Project Setup

The tutorial assumes your development environment already has PHP and Composer. Pull up the Laravel installer as a global Composer package, initialize a new Laravel project, and add the authentication boilerplate (with the `make:auth` command below):

```bash
composer global require laravel/installer
laravel new okta-laravel-auth && cd okta-laravel-auth
php artisan make:auth
php artisan serve
```

> Load http://localhost:8000/ and you'll see the default Laravel page with Login and Register links. However, these links will not work yet because you haven't run the database migrations yet, and the boilerplate code uses the local database for user registration/authorization.

Reviewer: back tics around http://localhost:8000/

In the next two sections, you'll set up a new Okta account/web application and connect Okta to the Laravel project so you can use it for authentication.

## Add Authentication to Laravel with Okta

> Before you proceed, you need to log into your Okta account (or [create a new one for free](https://developer.okta.com/signup/)) and create an OAuth application. You'll need to get a client ID and a client secret for your application.

Reviewer: OAuth > OIDC

> Start by going to the Applications menu item and clicking the 'Add Application' button:

Reviewer: 'Add Application' > **Add Application**

{% img blog/php-laravel-authentication/image2.png alt:"Add Application button" width:"300" %}{: .center-image }

> Select 'Web' and click 'Next'.

Reviewer: ```suggestion
Select **Web** and click **Next**.
```

> {% img blog/php-laravel-authentication/image7.png alt:"Select Web from the list of application types" width:"800" %}{: .center-image }

Reviewer: I'd suggest renaming the images to something more descriptive.

> Enter a title, set http://localhost:8000/ as the Base URI and http://localhost:8000/login/okta/callback as the Login Redirect URI, then click Done. You can leave the rest of the settings as they are.

Reviewer: ```suggestion
Enter a title, set `http://localhost:8000/` as the Base URI and `http://localhost:8000/login/okta/callback` as the Login Redirect URI, then click **Done**. You can leave the rest of the settings as they are.
```

> Copy the **Client ID** and **Client Secret** values from the application settings. Go to **Api > Authorization Servers**, and copy just the hostname part of the `Issuer URI` field (without the `/oauth2/default` part) - this is your Okta Base URL (it looks like `https://{{yourOktaDomain}}`).

Reviewer: ```suggestion
Copy the **Client ID** and **Client Secret** values from the application settings. Go to **API** > **Authorization Servers**, and copy just the hostname part of the `Issuer URI` field (without the `/oauth2/default` part) - this is your Okta Base URL (it looks like `https://{{yourOktaDomain}}`).
```

## Connect Okta to the Laravel Project

> Edit .env.example and add the following:

Reviewer: ```suggestion
Edit `.env.example` and add the following:
```

Reviewer: Should this be "rename `.env.example` to `.env`"?

```php
<?php

use Illuminate\Support\Facades\Schema;
use Illuminate\Database\Schema\Blueprint;
use Illuminate\Database\Migrations\Migration;

class CreateUsersTable extends Migration
{
    /**
     * Run the migrations.
     *
     * @return void
     */
    public function up()
    {
        Schema::create('users', function (Blueprint $table) {
            $table->bigIncrements('id');
            $table->string('email')->unique();
            $table->string('name');
            $table->text('token');
            $table->timestamps();
        });
    }

    /**
     * Reverse the migrations.
     *
     * @return void
     */
    public function down()
    {
        Schema::dropIfExists('users');
    }
}
```

Replace `app/User.php` with:

```php
<?php

namespace App;

use Illuminate\Notifications\Notifiable;
use Illuminate\Contracts\Auth\MustVerifyEmail;
use Illuminate\Foundation\Auth\User as Authenticatable;

class User extends Authenticatable
{
    use Notifiable;

    /**
     * The attributes that are mass assignable.
     *
     * @var array
     */
    protected $fillable = [
        'email', 'name', 'token'
    ];
}
```

> Let's use sqlite as the database engine (it's perfectly fine for this tutorial). Update `.env` and set the database connection to sqlite:

Reviewer: ```suggestion
You can use use sqlite as the database engine (it's perfectly fine for this tutorial). Update `.env` and set the database connection to sqlite:
```

        } else {
            // if the user already exists, just update the token:
            $localUser->token = $user->token;
            $localUser->save();
        }

        try {
            Auth::login($localUser);
        } catch (\Throwable $e) {
            return redirect('/login-okta');
        }
        
        return redirect('/home');
    }

    public function logout()
    {
        Auth::logout();
        return redirect('/');
    }
}
```

Modify the section of `resources/views/welcome.blade.php` that displays the top right menu:

```php
@if (Route::has('login'))
    <div class="top-right links">
        @auth
            <a href="{{ url('/home') }}">Home</a>
        @else
            <a href="{{ url('/login/okta') }}">Log in with Okta</a>
        @endauth
    </div>
@endif
```

Update `app/Http/Middleware/Authenticate.php`, replace `return route('login');` with `return route('login-okta');`

Run the development web server again with `php artisan serve` and load up `http://localhost:8000`. Click **Login with Okta** and after a successful login you will see the dashboard and the Okta user's name in the top right corner.

### Modify Home Page For Logged-in Users

> Open `resources/views/welcome.blade.php` and update the section that contains the 'Laravel' text so it shows some additional information is a user's logged in:

Reviewer: ```suggestion
Open `resources/views/welcome.blade.php` and update the section that contains the 'Laravel' text so it shows some additional information if a user is logged in:
```

```php
<div class="title m-b-md">
    Laravel
    @auth
        <p>You are logged in as {{ Auth::user()->name }}!</p>
    @endauth
</div>
```

> Go back to the home page at `http://localhost:8000/` and if you're logged in, you will see the text with your name in it.

Reviewer: ```suggestion
Go back to the home page at `http://localhost:8000/`. If you're logged in, you will see the text with your name in it.
```

## Add a Page For Logged-in Users Only

> Next, create a new route and controller method in HomeController for a new personal home page that will only be accessible to logged-in users.

Reviewer: ```suggestion
Next, create a new route and controller method in `HomeController` for a new personal home page that will only be accessible to logged-in users.
```

`routes/web.php`

```php
Route::get('/personal-home', 'HomeController@personal')->name('personal-home')->middleware('auth');
```

`app/Http/Controllers/HomeController.php`

```php
...
public function personal()
{
    return view('personal');
}
...
```

Create a new file `resources/views/personal.blade.php`

```php
@extends('layouts.app')

@section('content')
<div class="container">
    <div class="row justify-content-center">
        <div class="col-md-8">
            <div class="card">
                <div class="card-header">Personal Home Page</div>
                <div class="card-body">

>                     You are logged in as {{ Auth::user()->name }}! <br>

Reviewer: You need to wrap this code block with {% raw %} tags since it has double curlies in it. Might need it for other code blocks as well.

                    Your email address is {{ Auth::user()->email }}!
                </div>
            </div>
        </div>
    </div>
</div>
@endsection
```

Add a link to the home page that will only show to logged-in users in the Links section of `resources/views/welcome.blade.php`:

```php
<div class="links">
    @auth
        <a href="{{ route('personal-home') }}">Personal Home Page</a>
    @endauth
    <a href="https://laravel.com/docs">Docs</a>
    <a href="https://laracasts.com">Laracasts</a>
    <a href="https://laravel-news.com">News</a>
    <a href="https://blog.laravel.com">Blog</a>
    <a href="https://nova.laravel.com">Nova</a>
    <a href="https://forge.laravel.com">Forge</a>
    <a href="https://github.com/laravel/laravel">GitHub</a>
</div>
```

> Reload the home page and if you are logged in, you will see the 'Personal Home Page' link. Click on it to see the new page which displays your name and email address from Okta.

Reviewer: ```suggestion
Reload the home page and you will see the **Personal Home Page** link. Click on it to see the new page which displays your name and email address from Okta.
```

That's how easy it is to create a new Laravel application with secure user authentication via Okta!

> You can find the completed project [on GitHub](https://github.com/oktadeveloper/okta-php-laravel-auth).

Reviewer: I typically end the repo name with `-example`. Also, the README is a bit sparse. Keep in mind that the repo might show up under https://toolkit.okta.com/apps/ and its nice to have all the instructions in the README for cloning the repo and configuring the completed app to work with Okta. For example: https://github.com/oktadeveloper/okta-ionic4-jhipster-example

Author: I added some more details to the README, but didn't go as far as duplicating the detailed instructions on signing up for Okta and stuff.

## Learn More About Laravel, Authentication, and Okta

> If you find any issues, please add a comment below, and we'll do our best to help. If you liked this tutorial, you should[follow us on Twitter](https://twitter.com/oktadev). Follow us on [YouTube](https://www.youtube.com/c/oktadev) to find our screencasts and other videos.

Reviewer: ```suggestion
If you find any issues, please add a comment below, and we'll do our best to help. If you liked this tutorial, you should [follow us on Twitter](https://twitter.com/oktadev). [Follow us on YouTube](https://www.youtube.com/c/oktadev) to find our screencasts and other videos.
```

If you are interested in other Laravel tutorials that build a single-page application and use Okta for authentication, check out this series and build an example application with an Angular, React, or Vue frontend:

* [Build a Basic CRUD App with Laravel and Angular](/blog/2018/10/23/php-laravel-angular-crud-app)
* [Build a Basic CRUD App with Laravel and React](/blog/2018/12/06/crud-app-laravel-react)
* [Build a Basic CRUD App with Laravel and Vue](/blog/2019/01/15/crud-app-laravel-vue)
* [Build Simple Login in PHP](/blog/2018/12/28/simple-login-php)

> If you find any issues or have questions, please add a comment below! And if you liked this tutorial,[follow us on Twitter](https://twitter.com/oktadev) or our [YouTube channel](https://www.youtube.com/c/oktadev) where we publish screencasts and other videos.

Reviewer: Redundant - the same sentences exist above.
